Added initial support for a TensorFlow checkpoint file reader.

eaplatanios opened the pull request:
This PR adds initial support for a TensorFlow checkpoint file reader. I have not added tests yet because: (i) I don't know what's a good way to include resource files for swift tests, and (ii) I don't have a readily available small checkpoint. This is preliminary work while I'm trying to design a more standardized way of saving/loading checkpoints. Merging this now would unblock me and I can make subsequent PRs once I have a more concrete solution for the checkpointing issue.

P.S. I'm currently working on some ideas using `Codable` and `CodingKeys` to add compatibility with existing Python TF checkpoints.

rxwei commented:
`init(owning:)` is considered an implementation detail, and it should be hidden whenever possible. Instead, have you considered writing the the checkpoint reader within the library?

eaplatanios commented:
I can do that. I haven’t tested things yet so I wasn’t sure if you’d like
it to be merged. I will try to open a PR tomorrow. :)

On Wed, Oct 9, 2019 at 2:28 AM Richard Wei <notifications@github.com> wrote:

> init(owning:) is considered an implementation detail, and it should be
> hidden whenever possible. Instead, have you considered writing the the
> checkpoint reader within the library?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/swift-apis/pull/529?email_source=notifications&email_token=AAJ4EXGO22FPUFFKI26H743QNV2Y7A5CNFSM4I62TPEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAWY2WY#issuecomment-539856219>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJ4EXDXJ6HIMBCZZMAJNQDQNV2Y7ANCNFSM4I62TPEA>
> .
>

eaplatanios commented:
> `init(owning:)` is considered an implementation detail, and it should be hidden whenever possible. Instead, have you considered writing the the checkpoint reader within the library?

I changed this PR to add the checkpoint reader instead and updated the PR message.

eaplatanios commented:
> How do you load pretrained models o.O ?

Currently working in getting a hopefully nice solution. Will open another PR when I have one. :)

eaplatanios commented:
@rxwei on a separate note, at some point we should update the `checkOk` helper to throw an exception instead of a fatal error and mark all relevant functions that use it as `throws`.

eaplatanios commented:
@rxwei @dan-zheng gentle ping for a review.

dan-zheng commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+    @inlinable
+    public func load<Scalar: _TensorFlowDataTypeCompatible>(
+        tensorNamed name: String
+    ) -> ShapedArray<Scalar> {
```
I'm curious if there's some reason why `load` returns `ShapedArray<Scalar>` rather than `Tensor<Scalar>`?

eaplatanios replied:
This is just because it is stored in host memory anyway and would require an internal conversion to Tensor. In this case, I thought it’d be nice to let the user decide to convert if they want to. What do you think?

eaplatanios replied:
@dan-zheng do you think we can get this merged soon? I believe it is useful irrespective of how we design high-level serialization APIs as it allows us to read checkpoint files written by the TF Python API.

dan-zheng replied:
> This is just because it is stored in host memory anyway and would require an internal conversion to Tensor. In this case, I thought it’d be nice to let the user decide to convert if they want to. What do you think?

I see, thanks for explaining! I'm curious if you use `ShapedArray` much in your projects, in addition to `Tensor`?

eaplatanios replied:
Not really. The APIs for `ShapedArray` are very limited currently and I also mostly use GPUs for compute so I use `Tensor`s almost always. I just like to keep `ShapedArray` in APIs that could be used in both ways, so that it's up to the user to convert to a `Tensor`.

eaplatanios commented:
Thanks @dan-zheng ! I'll post examples and tests once I settle on a design. I have been mostly experimenting with loading BERT checkpoints for now and thinking about how the loading mechanism can become more modular. By the way, I have been unable to build the S4TF toolchain on my Mac lately. The build script hangs while building SwiftPM. This is the command it runs and the stacktrace I get when I kill the process:
```
/Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/debug/swiftc --driver-mode=swift -L /Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/lib/swift/pm/4_2 -lPackageDescription -swift-version 4.2 -I /Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/lib/swift/pm/4_2 -target x86_64-apple-macosx10.10 -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -package-description-version 4.2 /Users/eaplatanios/Development/GitHub/swift-source-temp/swiftpm/Package.swift -fileno 8

Traceback (most recent call last):
  File "/Users/eaplatanios/Development/GitHub/swift-source-temp/swiftpm/Utilities/bootstrap", line 1565, in <module>
    main()
  File "/Users/eaplatanios/Development/GitHub/swift-source-temp/swiftpm/Utilities/bootstrap", line 1334, in main
    result = subprocess.call(cmd, cwd=g_project_root)
  File "/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 172, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1099, in wait
    pid, sts = _eintr_retry_call(os.waitpid, self.pid, 0)
  File "/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 125, in _eintr_retry_call
    return func(*args)
```
I even let it run overnight and there was still a swift process fully consuming a single CPU core running all night. I'm using Catalina 10.15.1 Beta (19B77a) and Xcode 11.1 (11A1027). Do you have any idea what may be wrong?

eaplatanios commented:
@dan-zheng if I kill the background swift process itself, I get the following error:
```
/Users/eaplatanios/Development/GitHub/swift-source-temp/swiftpm: error: manifest parse error(s):
/Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftinterface:1252:16: error: referencing struct 'DifferentiableView' on 'Array' requires that 'Element' conform to '_Differentiable'
    if index < base.count {
               ^
/Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftinterface:1209:1: note: where 'Element' = 'Element'
extension Array where Element : Swift._Differentiable {
^
/Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftinterface:1253:14: error: referencing struct 'DifferentiableView' on 'Array' requires that 'Element' conform to '_Differentiable'
      return base[index]
             ^
/Users/eaplatanios/Development/GitHub/swift-source-temp/build/buildbot_osx/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftinterface:1209:1: note: where 'Element' = 'Element'
extension Array where Element : Swift._Differentiable {
^
--- bootstrap: error: build failed with exit status 1
```
Maybe this is not the best way to discuss this, but I'm wondering if there is a quick fix. It's weird that there is an error related to array differentiability when compiling SwiftPM.

eaplatanios commented:
The swift interface file referenced does not include at all the `& Differentiable` constraint shown [here](https://github.com/apple/swift/blob/7498390f620ee3b9e754eed6ddba58f5f61bf714/stdlib/public/core/Array.swift#L2003).

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+
+    /// Returns `true` if the checkpoint contains a tensor with the provided name.
+    @inlinable
+    public func contains(tensorNamed name: String) -> Bool {
```
`containsTensor(named:)`

eaplatanios replied:
Thanks Richard! I'll address all your comments in a separate PR since this was merged already.

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+
+    /// Returns the shape of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func shape(ofTensorNamed name: String) -> TensorShape {
```
`shapeOfTensor(named:)`

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+
+    /// Returns the data type of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func dataType(ofTensorNamed name: String) -> TensorDataType {
```
`TensorDataType` is an internal type that was never designed to be exposed to the user. It should have been named to `_TensorFlowDataType` according to the original design. A bigger problem is, in Swift for TensorFlow APIs, that we have never exposed the notion of "data type" to the user. The notion has always been "scalar type".

eaplatanios replied:
In principle, we could make this function internal. The issue though is that there is not way for a user to inspect the data types of tensors stored in a checkpoint. I could make it return `String` with the name of the data type instead. Do you think that would be better?

rxwei replied:
How about returning a metatype, `Any.Type`?

eaplatanios replied:
Sounds good. Will open a PR in the next few minutes.

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+
+    /// Loads and returns the value of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func load<Scalar: _TensorFlowDataTypeCompatible>(
```
`loadTensor(named:)`

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+    /// URL of the checkpoint file.
+    public let checkpointPath: URL
+
+    /// Number of tensors stored in the checkpoint.
```
Always add a definite article in doc comments that describe an entity.

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+    /// Number of tensors stored in the checkpoint.
+    public var tensorCount: Int { Int(TF_CheckpointReaderSize(handle)) }
+
+    /// Names of the tensors stored in the checkpoint.
```
Always add a definite article in doc comments that describe an entity.

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+    @usableFromInline internal let status: OpaquePointer
+    @usableFromInline internal let handle: OpaquePointer
+
+    /// URL of the checkpoint file.
```
Always add a definite article in doc comments that describe an entity.

rxwei commented on Sources/TensorFlow/Core/Serialization.swift:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /// Returns the shape of the tensor with the provided name stored in the checkpoint.
+    @inlinable
```
I'd suggest not doing `@inlinable` and `@usableFromInline` preemptively.

dan-zheng commented:
> By the way, I have been unable to build the S4TF toolchain on my Mac lately. The build script hangs while building SwiftPM. This is the command it runs and the stacktrace I get when I kill the process:

Thanks for sharing, and sorry for not catching the error! I just replicated the same hang while building SwiftPM, filed [TF-934](https://bugs.swift.org/browse/TF-934). I think the macOS toolchain issue started after we merged `swift-DEVELOPMENT-SNAPSHOT-2019-10-08-a` into `tensorflow` (https://github.com/apple/swift/pull/27751).

The team is looking into another merge from `swift-DEVELOPMENT-SNAPSHOT-2019-10-13-a`, perhaps that will naturally resolve the issue. If not, we will investigate further.

eaplatanios commented:
Great, thanks @dan-zheng! I'll also post there if I find a way to work around this issue.
